sqlhooks.go

```diff
@@ -110,6 +119,9 @@ func (conn *ExecerContext) ExecContext(ctx context.Context, query string, args [
 
 	results, err := conn.execContext(ctx, query, args)
 	if err != nil {
+		if h, ok := conn.hooks.(OnError); ok {
+			ctx, err = h.OnError(ctx, err, query, list...)
+		}
```

Comment:
I don't think you should override `err`. Otherwise `OnError` could supress errors that happen by returning nil for an error, which is likely won't most people will do.

I'd expect that if `OnError` did return a non-nil `err`, you return that error, otherwise you would return the original `err`.

Reply:
This is a good point, while error overriding/suppressing can generate confusions it might also be helpful if you want to bypass driver specific errors. I'm not sure here.

Reply:
If `err` is suppressed, hook must generate another `results`.
If both of them are `nil`, it is easy to panic. maybe I really need to change it to just like @keegancsmith said.

Reply:
I think of `Hooks` as something which doesn't influence behaviour, rather observes it. We currently can mutate what `ctx` is, but that is more to do with it being useful to store some "request" data in it that we can then access in `After`. I think being able to cancel out err will lead to API which is easy to make a mistake in. Especially since we could then return `nil, nil`.

Reply:
@keegancsmith I agree, but do you think it might be useful to suppress or modify errors?
I'm not talking about implementing it in this `Hook` or not, but if it's useful we might need to come up with some  way to do it.